mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/query/parsing/nativequery/impl/NativeQueryParser.java

```diff
@@ -57,301 +67,300 @@
-				") "
-		);
+	public Rule PathExpression() {
+		return Sequence( IdentiferInPath(), ZeroOrMore( Separator(), IdentiferInPath() ) );
 	}
 
-	public Rule UpdateMany() {
-		return Sequence(
-				Separator(),
-				"updateMany ",
-				"( ",
-				JsonObject(), builder.setCriteria( match() ), ", ",
-				JsonObject(), builder.setUpdateOrInsert( match() ),
-				Optional( Sequence( ", ", JsonObject(), builder.setOptions( match() ) ) ),
-				") "
-		);
+	public Rule IdentiferInPath() {
+		return Sequence( Identifier(), cliQueryInfo.pathExpression.add( match().trim() ) );
 	}
 
-	public Rule ReplaceOne() {
+	@SuppressSubnodes
+	public Rule Identifier() {
 		return Sequence(
-				Separator(),
-				"replaceOne ",
-				"( ",
-				JsonObject(), builder.setCriteria( match() ), ", ",
-				JsonObject(), builder.setUpdateOrInsert( match() ),
-				Optional( Sequence( ", ", JsonObject(), builder.setOptions( match() ) ) ),
-				") "
+				FirstOf( SpecialIdentifierCharacter(), Letter()),
+				ZeroOrMore( FirstOf( SpecialIdentifierCharacter(), Letter(), Digit() )),
+				ZeroOrMore( WhiteSpace() )
 		);
 	}
 
-	public Rule Aggregate() {
-		return Sequence( Separator(), "aggregate ", "( ", AggregateArray(), ") " );
+	public Rule Arguments() {
+		return Optional( Sequence( Argument(), ZeroOrMore( ", ", Argument() ) ) );
 	}
 
-	public Rule AggregateArray() {
+	public Rule Argument() {
 		return Sequence(
-				"[ ",
-					Sequence(
-						AggregateObject(),
-						ZeroOrMore( Sequence( ", ", AggregateObject() ) ) ),
-				"] " );
-	}
-
-	public Rule AggregateObject() {
-		return Sequence(
-				ZeroOrMore( WhiteSpace() ).skipNode(),
-				"{ ", AggregatePair(), "} " );
+				Value(),
+				cliQueryInfo.arguments.add( match() )
+				);
 	}
 
-	public Rule AggregatePair() {
-		return Sequence(
-				JsonString(), builder.push( currentIndex(), match() ),
-				": ",
-				Value(), builder.addPipeline( builder.pop(), match() ) );
-	}
-
-	public Rule Count() {
-		return Sequence(
-				Separator(),
-				"count ",
-				"( ",
-				Optional( Sequence( JsonComposite(), builder.setCriteria( match() ) ) ),
-				") "
-		);
+	@SuppressNode
+	public Rule Separator() {
+		return Sequence( ZeroOrMore( WhiteSpace() ), ". " );
 	}
 
-	public Rule Distinct() {
-		return Sequence(
-				Separator(),
-				"distinct ",
-				"( ",
-				Sequence( JsonString(), builder.setDistinctFieldName( JSON.parse( match() ).toString() ) ),
-				Optional( Sequence( ", ", JsonObject(), builder.setCriteria( match() ) ) ),
-				Optional( Sequence( ", ", JsonObject(), builder.setOptions( match() ) ) ),
-				") "
-		);
-	}
 
-	public Rule MapReduce() {
-		return Sequence(
-				Separator(),
-				"mapReduce ",
-				"( ",
-				Sequence( JsonString(), builder.setMapFunction( JSON.parse( match() ).toString() ) ),
-				Sequence( ", ", JsonString(), builder.setReduceFunction( JSON.parse( match() ).toString() ) ),
-				Optional( Sequence( ", ", JsonObject(), builder.setOptions( match() ) ) ),
-				") "
```

Comment:
So what I'm wondering if that maybe we could have something like:
```
FirstOf(
    the valid sequence of parameters (i.e. what I pasted below),
    throw a parse error
)
```
instead of
```
			       "( ",
				Sequence( JsonString(), builder.setMapFunction( JSON.parse( match() ).toString() ) ),
				Sequence( ", ", JsonString(), builder.setReduceFunction( JSON.parse( match() ).toString() ) ),
				Optional( Sequence( ", ", JsonObject(), builder.setOptions( match() ) ) ),
				") "
```

Reply:
I think that the more generic `Query` rule should throw errors, this way the parser can first identify the general structure of a query and fail only when it has all the necessary information.